lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,10 +12,92 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
+  ExprMap m_cache;
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  }
+
+  bool isVar(Expr exp) { return bind::isBoolVar(exp) || bind::isIntVar(exp); }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
+
+  Expr inferTypeITE(Expr exp) {
+    Expr boolSort = sort::boolTy(exp->efac());
+
+    // ite(a,b,c) : a is bool type, b and c are the same type
+    if (exp->arity() == 3 && m_cache.at(exp->arg(0)) == boolSort &&
+        (m_cache.at(exp->arg(1)) == m_cache.at(exp->arg(2))))
+      return boolSort;
+    else
+      return Expr();
+  }
+
+  // ensures the expression has the correct number of children and all children
+  // are bool types
+  Expr boolCheckChildren(Expr exp, bool (*checkNumChildren)(int)) {
+    auto isBool = [this](Expr exp) {
+      return this->m_cache.at(exp) == sort::boolTy(exp->efac());
+    };
+
+    if (checkNumChildren(exp->arity()) &&
+        std::all_of(exp->args_begin(), exp->args_end(), isBool))
+      return sort::boolTy(exp->efac());
+    else
+      return Expr();
+  }
+
+  Expr inferTypeBoolBinary(Expr exp) {
+    auto checkNumChildren = [](int numChildren) -> bool {
+      return numChildren == 2;
+    };
+    return boolCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeBoolUnary(Expr exp) {
+    auto checkNumChildren = [](int numChildren) -> bool {
+      return numChildren == 1;
+    };
+    return boolCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeBoolNary(Expr exp) {
+    auto checkNumChildren = [](int numChildren) -> bool {
+      return numChildren >= 2;
+    };
+    return boolCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferType(Expr exp /*,TypeChecker & ty*/) {
+    if (isOpX<TRUE>(exp) || isOpX<FALSE>(exp))
+      return sort::boolTy(exp->efac());
+    else if (bind::isBoolVar(exp) || bind::isBoolConst(exp))
+      return sort::boolTy(exp->efac());
+    else if (isOpX<AND>(exp) || isOpX<OR>(exp) || isOpX<XOR>(exp))
+      return inferTypeBoolNary(exp);
+    else if (isOpX<NEG>(exp))
+      return inferTypeBoolUnary(exp);
+    else if (isOpX<IMPL>(exp) || isOpX<IFF>(exp))
+      return inferTypeBoolBinary(exp);
+    else if (isOpX<ITE>(exp))
+      return inferTypeITE(exp);
+    else if (bind::isIntVar(exp) || bind::isIntConst(exp))
+      return sort::intTy(exp->efac());
+
+    return sort::boolTy(exp->efac());
+  }
 
   /// Called after children have been visited
   Expr postVisit(Expr exp) {
     LOG("tc", llvm::errs() << "post visiting expression: " << *exp << "\n";);
+
+    Expr type = inferType(exp);
+
+    m_cache.insert({exp, type});
+    if (!type)
```

Comment:
need a good way to bailout. You can't just terminate the visitor, but you can ensure that pre-visit returns false from this point on. 

Also need a strategy to remember what was the failure and a way to report this to clients.